rafthttp: support sending v3 snapshot

yichengq opened the pull request:
After sending raft snapshot message as before, sender will wait for receiver to request the full snapshot and send it to the receiver.
Only after sender sends out the full snapshot, it thinks that snapshot is sent successfully.
Only after receiver receives the full snapshot, it acknowledges the snapshot is received and process the raft snapshot message.

Compared to using stream, this method handles the case that how remote struct receives snapshot well, let external have the control about how to read/write snapshot, and doesn't require to maintain a long-term connection for rare snapshot message.

I have tested it manually and it works well.

for #3549

xiang90 commented on rafthttp/snapshot_hub.go:
```diff
@@ -0,0 +1,170 @@
+
+	// protect fields below
+	mu sync.Mutex
+	// whether it has registered snapshot that is going to be written out or closed
```
Why `that is going to be written out or closed` is important to this variable? I cannot see a clear relationship between this boolean variable and the comment.

xiang90 commented on rafthttp/snapshot_hub.go:
```diff
@@ -0,0 +1,170 @@
+	mu sync.Mutex
+	// whether it has registered snapshot that is going to be written out or closed
+	hasSnap   bool
+	snapTo    types.ID
```
what is snapTo? This is unclear.

xiang90 commented on rafthttp/snapshot_hub.go:
```diff
@@ -0,0 +1,170 @@
+	hasSnap   bool
+	snapTo    types.ID
+	snapIndex uint64
+	// when it fires, it closes the registered snapshot
```
who is the second `it`? why the timer can close the snapshot?

xiang90 replied:
// when the timer expires, it will call the func attached via AfterFunc to close the registered snapshot.

xiang90 commented on rafthttp/snapshot_hub.go:
```diff
@@ -0,0 +1,170 @@
+
+// registerSnap registers snapshot in snapshot hub.
+// The registered snapshot waits to be written out through writeSnapTo func.
+// If it exceeds wait timeout, the snapshot is closed and deregistered.
```
what is the difference from closed and deregistered?

yichengq replied:
`If the snapshot exceeds wait timeout, it is closed and snapshot hub deregisters it.`

xiang90 commented on rafthttp/snapshot_hub.go:
```diff
@@ -0,0 +1,170 @@
+	snapTo    types.ID
+	snapIndex uint64
+	// when it fires, it closes the registered snapshot
+	snapClose *time.Timer
```
snapCloseTimer

xiang90 commented on rafthttp/snapshot_hub.go:
```diff
@@ -0,0 +1,170 @@
+		return fmt.Errorf("failed to find available snapshot")
+	}
+	if s.snapIndex != index {
+		return fmt.Errorf("failed to find wanted snapshot")
```
can we be more specific with the error? failed to find the snapshot with index %d, index

xiang90 commented on rafthttp/snapshot_hub.go:
```diff
@@ -0,0 +1,170 @@
+	if err != nil {
+		plog.Infof("snapshot [index: %d, to: %s] failed to be sent out (%v)", index, s.snapTo, err)
+		s.r.ReportSnapshot(uint64(s.snapTo), raft.SnapshotFailure)
+	} else {
```
remove the else. return in err!=nil case.

xiang90 commented on rafthttp/snapshot_hub.go:
```diff
@@ -0,0 +1,170 @@
+		plog.Infof("snapshot [index: %d, to: %s] failed to be sent out (%v)", index, s.snapTo, err)
+		s.r.ReportSnapshot(uint64(s.snapTo), raft.SnapshotFailure)
+	} else {
+		plog.Infof("snapshot [index: %d, to: %s] succeeded to be sent out", index, s.snapTo)
```
remove `to be send out` or `snapshot sent out successfully`

xiang90 commented on rafthttp/snapshot_hub.go:
```diff
@@ -0,0 +1,170 @@
+func (s *snapshotHub) requestSnap(remote types.ID, index uint64) (io.ReadCloser, error) {
+	up, ok := s.urlPickers[remote]
+	if !ok {
+		return nil, fmt.Errorf("cannot find url picker")
```
what do you mean by cannot find a picker? this will be logged out as warning. I do not expect our user to understand the internal of rafthttp.

xiang90 commented on rafthttp/snapshot_hub.go:
```diff
@@ -0,0 +1,170 @@
+	uu := u
+	uu.Path = RaftSnapPrefix
+	req, err := http.NewRequest("GET", uu.String(), nil)
+	if err != nil {
```
when newreuqest will return an error?

yichengq replied:
I don't think it will return error. let me panic it here.

xiang90 commented on rafthttp/snapshot_hub.go:
```diff
@@ -0,0 +1,170 @@
+		return resp.Body, nil
+	default:
+		resp.Body.Close()
+		return nil, fmt.Errorf("unexpected http status %s", http.StatusText(resp.StatusCode))
```
shall we also mark unreachable in this case?

yichengq replied:
The url is reachable but doesn't work, which indicates that users provide wrong-functioning peer urls. When this happens, I think it should keep printing errors and users should notice about this and corrects the peer urls.

xiang90 replied:
Wait... I totally understand it is because user provided a wrong one. And we SHOULD log the error. However my question was `shall we also mark unreachable in this case?`.

yichengq replied:
Maybe yes. It is good to work as long as it is possible.

xiang90 replied:
> It is good to work as long as it is possible.

This contradicts with what you said in the previous comment...

xiang90 commented on etcdserver/snapshot_store.go:
```diff
@@ -83,28 +83,45 @@ func (ss *snapshotStore) getSnap() (*snapshot, error) {
+}
+
+// ReadSnapFrom reads the snapshot from the given reader, and saves it
+// into the file.
```
what is the file?

xiang90 commented on etcdserver/snapshot_store.go:
```diff
@@ -83,28 +83,45 @@ func (ss *snapshotStore) getSnap() (*snapshot, error) {
+
+// ReadSnapFrom reads the snapshot from the given reader, and saves it
+// into the file.
+func (ss *snapshotStore) ReadSnapFrom(r io.Reader, index uint64) error {
```
ReadAndSaveSnapFrom

xiang90 replied:
Or just SaveSnap

yichengq replied:
will use `SaveSnapFrom`.

xiang90 commented on rafthttp/http.go:
```diff
@@ -214,6 +229,25 @@ func (h *streamHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	index, err := strconv.ParseUint(r.Header.Get("X-Snapshot-Index"), 10, 64)
+	if err != nil {
+		plog.Errorf("failed to parse X-Snapshot-Index header %q (%v)", index, err)
+		http.Error(w, "invalid index", http.StatusNotFound)
```
http.StatusNotFound -> http.StatusBadRequest?

yichengq replied:
agree.

xiang90 commented on rafthttp/transport.go:
```diff
@@ -85,17 +93,21 @@ type transport struct {
+	term       uint64               // the latest term that has been observed
+	remotes    map[types.ID]*remote // remotes map that helps newly joined member to catch up
+	peers      map[types.ID]Peer    // peers map
+	urlPickers map[types.ID]*urlPicker
```
Why do we need to have the picker here? Cannot we access the peer first then picker?

yichengq replied:
I don't fully understand this sentence: `Cannot we access the peer first then picker?`

xiang90 replied:
I do not think we need to have the url picker. we can get the peer first. each peer has its own picker, no?

yichengq replied:
> each peer has its own picker

yes

I use urlPickers because fetching it is inconvenient:
1. it exists in peer struct in most cases, and otherwise it exists in remote struct(which happens for newly-joined member)
2. it needs to pass peer and remote map into snapshotHub, which sounds heavy

I may miss the importance of some concepts. Do you perfer to fetch url picker from peer/remote map?

xiang90 replied:
I think you still do not answer my original question or my question is not clear enough. My question is why do we need to have a picker map here in `transporter`.

xiang90 commented on rafthttp/snapshot_hub.go:
```diff
@@ -0,0 +1,170 @@
+	}
+}
+
+type errWriter struct {
```
add a comment for this please.

yichengq commented:
updated according to comments. Main change:
1. split snapshotHub into snapshotBuffer and snapshotProcessor to make each struct more straightforward and readable.
2. use urlPickerGetter interface to get urlPicker. `transport` struct is the one that manages peers and remotes, and it is over-complicated to add a new layer between them IMO. Use an interface could address this well.
3. rename ReadSnapFrom to SaveSnapFrom
4. errWriter is moved to util.go

yichengq commented:
Have added missing comments around snapshotBuffer and snapshotProcessor.

xiang90 commented on rafthttp/snapshot.go:
```diff
@@ -0,0 +1,188 @@
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+// timeout that waits snapshot to be called to write out
```
This comment is very hard to understand. It is harder to understand than just reading the code actually.

xiang90 commented on rafthttp/snapshot.go:
```diff
@@ -0,0 +1,188 @@
+	}
+}
+
+// snapshotBuffer represents a buffer of snapshot with write method.
```
I cannot understand the comment. :(
What is the responsibility of the struct?

yichengq replied:
`snapshotBuffer represents a buffer of snapshot. Snapshot could be added in the buffer, and writes to Writer when calling writeSnapTo method.` what about this one?

xiang90 replied:
I failed to see where we add the snapshot to the buffer This is a simple wrap around snapStore. All it does is to mark a snapshot. I have no idea why it is called buffer at the first place.

xiang90 commented on rafthttp/snapshot.go:
```diff
@@ -0,0 +1,188 @@
+	snapCloseTimer *time.Timer
+}
+
+// addSnap adds snapshot into snapshot buffer.
```
where does it add the snapshot?

yichengq replied:
At the place that  the raft snapshot message is going to be sent out.

xiang90 replied:
still cannot find. can you point me to the exact line in this function?

yichengq replied:
It adds the snapshot at https://github.com/coreos/etcd/pull/3648/files#diff-15a30f78ad76cc19daa724a361c77fc2R86 .

xiang90 commented on rafthttp/snapshot.go:
```diff
@@ -0,0 +1,188 @@
+	return nil
+}
+
+// snapshotProcessor processes incoming raft snapshot message.
```
// snapshotProcessor is used to process raft snapshot.
// When receiving a snapMsg, the transport layer should send the snapMsg to snapshotProcessor.
// snapshotProcessor helps to persist the snapshot and then ask raft state machine to further process 
// the snapMsg.

xiang90 commented on rafthttp/snapshot.go:
```diff
@@ -0,0 +1,188 @@
+// it lets snapshot store read out the snapshot. After the full snapshot is
+// received, it lets raft process the raft snapshot message.
+func (s *snapshotProcessor) processSnap(m raftpb.Message) {
+	remote := types.ID(m.From)
```
panic if message type is not snapMsg?

yichengq replied:
ok

xiang90 commented on rafthttp/snapshot.go:
```diff
@@ -0,0 +1,188 @@
+// First, it sends requests to get full snapshot for raft snapshot message. Then
+// it lets snapshot store read out the snapshot. After the full snapshot is
+// received, it lets raft process the raft snapshot message.
+func (s *snapshotProcessor) processSnap(m raftpb.Message) {
```
Let's rename this to `process`.

xiang90 commented on rafthttp/snapshot.go:
```diff
@@ -0,0 +1,188 @@
+	snapst   SnapshotStore
+}
+
+// processSnap processes the given raft snapshot message.
```
// process processes the given msgSnap.
// First, it requests the corresponding snapshot data from the sender of msgSnap.
// Then it persist the snapshot data into SnapshotStore. And finally, it sends the 
// msgSnap to raft for further processing.

xiang90 commented on rafthttp/snapshot.go:
```diff
@@ -0,0 +1,188 @@
+	}
+}
+
+// getSnapFrom gets full snapshot with the given index from the remote member.
```
// getSnapFrom requests the snapshot data with given index from the remote.
// It returns a ReadCloser for the snapshot data when succeeds. Otherwise, an
// error will be returned.

xiang90 commented on rafthttp/snapshot.go:
```diff
@@ -0,0 +1,188 @@
+func (s *snapshotProcessor) getSnapFrom(remote types.ID, index uint64) (io.ReadCloser, error) {
+	up := s.upGetter.getURLPicker(remote)
+	if up == nil {
+		return nil, fmt.Errorf("unknown member %s", remote)
```
when will this happen? how should our user deal with this error?

xiang90 commented on rafthttp/transport.go:
```diff
@@ -37,6 +38,13 @@ type Raft interface {
 	ReportSnapshot(id uint64, status raft.SnapshotStatus)
 }
 
+type SnapshotStore interface {
```
add comment for snapshotStore.

xiang90 replied:
I think you told me that you will put this into etcdserver/snapshotStore or something like that.

yichengq replied:
will add.

I originally mentioned to avoid snapshot concept is defined in rafthttp package, which is totally weird. Now it turns out that the interface does not need to understand what the snapshot is anymore, and it just provides io.Reader and io.Writer to read/write snapshot. So it is fine to put the interface here.

xiang90 commented on rafthttp/transport.go:
```diff
@@ -37,6 +38,13 @@ type Raft interface {
 }
 
+type SnapshotStore interface {
+	// WriteSnapTo writes the latest snapshot into the given writer.
```
what does latest mean?

yichengq replied:
it just makes comment unclear, will delete

xiang90 commented on rafthttp/util.go:
```diff
@@ -99,3 +99,10 @@ func checkVersionCompability(name string, server, minCluster *semver.Version) er
 }
+
+// errWriter always returns error when calling Write.
+type errWriter struct {
```
this should be put into pkg/util not util.go in rafthttp.

xiang90 commented on rafthttp/snapshot.go:
```diff
@@ -0,0 +1,188 @@
+		}
+		s.hasSnap = false
+		// close the snapshot in the store
+		s.snapst.WriteSnapTo(&errWriter{errors.New("cancelled")})
```
can we have a better error message than just `cancelled`?

yichengq replied:
ok

yichengq commented:
Updated based on discussion. The comment should be improved a lot.

xiang90 commented on etcdserver/snapshot_store.go:
```diff
@@ -20,30 +20,46 @@ import (
-
-func (s *snapshot) writeTo(w io.Writer) (n int64, err error) { return s.kv.WriteTo(w) }
+func (s *snapshot) WriteTo(w io.Writer) (n int64, err error) {
+	defer func() {
```
hmm... can we just put this after s.kv.writeto?

yichengq replied:
we could, and it should be much readable.

xiang90 commented on etcdserver/snapshot_store.go:
```diff
@@ -67,7 +83,7 @@ func newSnapshotStore(dir string, kv dstorage.KV) *snapshotStore {
 // If there is no available snapshot, ErrSnapshotTemporarilyUnavaliable will be returned.
 func (ss *snapshotStore) getSnap() (*snapshot, error) {
-	if ss.snap != nil {
+	if ss.snap != nil && !ss.snap.isClosed() {
```
i do not quite understand this. should it be && ss.snap.isClosed() instead?

yichengq replied:
the code is correct, but it is not straightforward when reading the code. If snapshotStore has some open snapshot still, it cannot request new snapshot and return no available snapshot. I suppose that adding some comments could help to understand.

xiang90 commented on etcdserver/snapshot_store.go:
```diff
@@ -80,26 +96,28 @@ func (ss *snapshotStore) getSnap() (*snapshot, error) {
 		r:  raftsnap,
 		kv: kvsnap,
 	}
+	ss.tr.SnapshotReady(raftsnap.Metadata.Index, ss.snap)
```
add a comment for this.

xiang90 commented on rafthttp/http.go:
```diff
@@ -107,7 +120,17 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	// process raft message
+	if h.v3demo && isMsgSnap(m) {
+		peerURLs, err := types.NewURLs(strings.Split(r.Header.Get("X-Request-Peer-URLs"), ","))
```
please add a comment on this.

xiang90 commented on rafthttp/http.go:
```diff
@@ -214,6 +237,40 @@ func (h *streamHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+// snapHandler handles the request for snapshot data.
+// When it receives a request, it parses out the index of target snapshot
+// and retrieves it from snapKeeper. Then it sends snapshot data to
+// the client, and reports send result to raft.
```
who is the client?

yichengq replied:
the sender of the request. i use client because this is a HTTP handler, and client is a known word for it.

xiang90 replied:
this is unclear. please make this clear in the comment.

xiang90 replied:
report snapshot sent status

xiang90 commented on rafthttp/http.go:
```diff
@@ -107,7 +120,17 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
-	if err := h.r.Process(context.TODO(), m); err != nil {
+
+	// process raft message
```
process raft snapshot message

xiang90 commented on rafthttp/http.go:
```diff
@@ -214,6 +237,40 @@ func (h *streamHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	plog.Infof("snapshot [index: %d, to: %s] sent out successfully", index, to)
+	h.r.ReportSnapshot(uint64(to), raft.SnapshotFinish)
+	return
+
```
remove extra empty space.

xiang90 commented on rafthttp/http.go:
```diff
@@ -214,6 +237,40 @@ func (h *streamHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+}
+
+func (h *snapHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	index, err := strconv.ParseUint(r.Header.Get("X-Snapshot-Index"), 10, 64)
```
X-Raft-Snapshot-Index. Please add prefix.

yichengq replied:
ok

xiang90 commented on rafthttp/http.go:
```diff
@@ -214,6 +237,40 @@ func (h *streamHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	index, err := strconv.ParseUint(r.Header.Get("X-Snapshot-Index"), 10, 64)
+	if err != nil {
+		plog.Errorf("failed to parse X-Snapshot-Index header %q (%v)", index, err)
+		http.Error(w, "invalid index", http.StatusBadRequest)
```
cannot found X-Raft-Snapshot-Index header. please be specific.

yichengq replied:
ok

xiang90 commented on rafthttp/snapshot.go:
```diff
@@ -0,0 +1,230 @@
+// be sent out, and snapHandler waits remote member requesting
+// snapshot data. The process needs to take two maximal RTT, which is 10s.
+// So we set 15s as the timeout.
+const snapshotAutoReleaseTimeout = 15 * time.Second
```
why this is a const rather than a var if this has something to do with RTT?

yichengq replied:
I use const because this is simpler. I know that a short snapshotAutoReleaseTimeout could help etcd to attempt snapshot send fast, but 15s is good enough considering snapshot send is a rare case.

xiang90 replied:
No. It blocks other snapshot sending.

yichengq replied:
ok. I will set it based on RTT then.

xiang90 commented on rafthttp/snapshot.go:
```diff
@@ -0,0 +1,230 @@
+func (s *snapshotKeeper) autoRelease(index uint64, timeout time.Duration) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.index != index {
```
why do we even need to have these checking?

yichengq replied:
to detect the potential bug early. If this happens, it indicates that snapshot data cannot match the index, which could let follower recover to mismatch storage.

xiang90 replied:
I do not feel good with so many panics.

xiang90 commented on rafthttp/snapshot.go:
```diff
@@ -0,0 +1,230 @@
+// If the snapshot is released automatically, snapshotKeeper also closes
+// the snapshot, and reports snapshot failure to raft.
+// If there is no available snapshot that has attached remote, it panics.
+func (s *snapshotKeeper) autoRelease(index uint64, timeout time.Duration) {
```
why do you need to separate this out from keep?

yichengq replied:
the closeTimer has to know the remote when it reports snapshot result to raft.

xiang90 commented on rafthttp/snapshot.go:
```diff
@@ -0,0 +1,230 @@
+	}
+}
+
+// getSnapFrom requests the given url for snapshot data at the given index.
```
requests the snapshot... from given url

yichengq commented:
Addressed all the comments.

For the too many panics stuff, I converge the index check into one place, and remove some unnecessary check.

yichengq commented:
Close for #3666
